libs/execution_context/ebpf_program.c

```diff
@@ -1175,3 +1213,180 @@ ebpf_program_create_and_initialize(
+ * @return EBPF_SUCCESS the program info hash matches.
+ * @return EBPF_INVALID_ARGUMENT the program info hash does not match.
+ */
+static ebpf_result_t
+_ebpf_program_verify_program_info_hash(_In_ const ebpf_program_t* program)
+{
+    EBPF_LOG_ENTRY();
+    ebpf_result_t result;
+    ebpf_cryptographic_hash_t* cryptographic_hash = NULL;
+    ebpf_helper_id_to_index_t* helper_id_to_index = NULL;
+    const ebpf_program_info_t* program_info = NULL;
+
+    result = ebpf_program_get_program_info(program, &program_info);
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    helper_id_to_index =
+        (ebpf_helper_id_to_index_t*)ebpf_allocate(program_info->count_of_helpers * sizeof(ebpf_helper_id_to_index_t));
+    if (helper_id_to_index == NULL) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+
+    for (uint32_t index = 0; index < program_info->count_of_helpers; index++) {
+        helper_id_to_index[index].helper_id = program_info->helper_prototype[index].helper_id;
+        helper_id_to_index[index].index = index;
+    }
+
+    // Sort helper_id_to_index by helper_id.
+    qsort(
+        helper_id_to_index,
+        program_info->count_of_helpers,
+        sizeof(ebpf_helper_id_to_index_t),
+        _ebpf_helper_id_to_index_compare);
+
+    result = ebpf_cryptographic_hash_create(EBPF_HASH_ALGORITHM, &cryptographic_hash);
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    // Hash is performed over the following fields:
+    // 1. Program type name.
+    // 2. Context descriptor.
+    // 3. Program type.
+    // 4. BPF program type.
+    // 5. Is_privileged flag.
+    // 6. Count of helpers.
+    // 7. For each helper (in helper id order).
+    //   a. Helper id.
+    //   b. Helper name.
+    //   c. Helper return type.
+    //   d. Helper argument types.
+
+    result = ebpf_cryptographic_hash_append(
+        cryptographic_hash,
+        (uint8_t*)program_info->program_type_descriptor.name,
+        strlen(program_info->program_type_descriptor.name));
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    result = ebpf_cryptographic_hash_append(
+        cryptographic_hash,
+        (uint8_t*)program_info->program_type_descriptor.context_descriptor,
+        sizeof(ebpf_context_descriptor_t));
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    result = ebpf_cryptographic_hash_append(
+        cryptographic_hash, (uint8_t*)&program_info->program_type_descriptor.program_type, sizeof(ebpf_program_type_t));
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    result = ebpf_cryptographic_hash_append(
+        cryptographic_hash, (uint8_t*)&program_info->program_type_descriptor.bpf_prog_type, sizeof(uint32_t));
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    result = ebpf_cryptographic_hash_append(
+        cryptographic_hash, (uint8_t*)&program_info->program_type_descriptor.is_privileged, sizeof(char));
```

Comment:
```suggestion
        cryptographic_hash, (uint8_t*)&program_info->program_type_descriptor.is_privileged, sizeof(program_info->program_type_descriptor.is_privileged));
```
Seems safest to do sizeof the actual field in case the type changes in the future.  Same for line 1326, 1334, etc.

Reply:
Agreed. Added a macro to make it easier to do the right thing (it accepts a reference to the value to hash and then internally uses sizeof, similar to the template [hash_t::append_byte_range](https://github.com/microsoft/ebpf-for-windows/blob/92fea6d08e2151f43ee3e817fd26e8b23411496e/tools/bpf2c/bpf2c.cpp#L46)).